### DISCUSSION TITLE
Does something like `ConvertTo-TreeTable` exist?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#21343] [Q&A] [answered]
If not evident, by “tree table”, I refer to what https://stackoverflow.com/q/77935915/9731176 describes. Specifically, say I have some JSON5 data like:

```JSON5
{
	"Candidate":
	{
		"Name": {
			"3rd": "Surname",
			"1st": "Forename",
			"2nd": "Middle name"
		},
		"Date of Birth": "+1582-10-15",
		"E-mail Address": "mailTo:address@domain.TLD",
		"Mobile Telephone Number": "tel:+00-0000-000000",
		"Blood Group": "" // Unknown
	},
	"Mother":
	{
		"Name": {
			"3rd": "Surname",
			"1st": "Forename",
			"2nd": "Middle name"
		},
		"Date of Birth": "+1582-10-15",
		"E-mail Address": "mailTo:address@domain.TLD",
		"Mobile Telephone Number": "tel:+00-0000-000000",
	},
	"Father":
	{
		"Name": {
			"3rd": "Surname",
			"1st": "Forename",
			"2nd": "Middle name"
		},
		"Date of Birth": "+1582-10-15",
		"E-mail Address": "mailTo:address@domain.TLD",
		"Mobile Telephone Number": "tel:+00-0000-000000",
	},
}
```

If I import this like

```PS1
$Values = @'
{
	"Candidate":
	{
		"Name": {
			"3rd": "Surname",
			"1st": "Forename",
			"2nd": "Middle name"
		},
		"Date of Birth": "+1582-10-15",
		"E-mail Address": "mailTo:address@domain.TLD",
		"Mobile Telephone Number": "tel:+00-0000-000000",
		"Blood Group": "" // Unknown
	},
	"Mother":
	{
		"Name": {
			"3rd": "Surname",
			"1st": "Forename",
			"2nd": "Middle name"
		},
		"Date of Birth": "+1582-10-15",
		"E-mail Address": "mailTo:address@domain.TLD",
		"Mobile Telephone Number": "tel:+00-0000-000000",
	},
	"Father":
	{
		"Name": {
			"3rd": "Surname",
			"1st": "Forename",
			"2nd": "Middle name"
		},
		"Date of Birth": "+1582-10-15",
		"E-mail Address": "mailTo:address@domain.TLD",
		"Mobile Telephone Number": "tel:+00-0000-000000",
	},
}
'@
$Values | ConvertFrom-JSON
$Values | ConvertFrom-JSON | Select-Object -ExpandProperty 'Candidate'
```

This gives me firstly a useless table and then a table of a very small amount of attributes:

<blockQuote>

```log
PS /home/RokeJulianLockhart> $Values = @'
>> {
>> "Candidate":
>> {
>> "Name": {
>> "3rd": "Surname",
>> "1st": "Forename",
>> "2nd": "Middle name"
>> },
>> "Date of Birth": "+1582-10-15",
>> "E-mail Address": "mailTo:address@domain.TLD",
>> "Mobile Telephone Number": "tel:+00-0000-000000",
>> "Blood Group": "" // Unknown
>> },
>> "Mother":
>> {
>> "Name": {
>> "3rd": "Surname",
>> "1st": "Forename",
>> "2nd": "Middle name"
>> },
>> "Date of Birth": "+1582-10-15",
>> "E-mail Address": "mailTo:address@domain.TLD",
>> "Mobile Telephone Number": "tel:+00-0000-000000",
>> },
>> "Father":
>> {
>> "Name": {
>> "3rd": "Surname",
>> "1st": "Forename",
>> "2nd": "Middle name"
>> },
>> "Date of Birth": "+1582-10-15",
>> "E-mail Address": "mailTo:address@domain.TLD",
>> "Mobile Telephone Number": "tel:+00-0000-000000",
>> },
>> }
>> '@
PS /home/RokeJulianLockhart> $Values | ConvertFrom-JSON

Candidate                                                                                                                                Mother                                                                                                                     Father
---------                                                                                                                                ------                                                                                                                     ------
@{Name=; Date of Birth=+1582-10-15; E-mail Address=mailTo:address@domain.TLD; Mobile Telephone Number=tel:+00-0000-000000; Blood Group=} @{Name=; Date of Birth=+1582-10-15; E-mail Address=mailTo:address@domain.TLD; Mobile Telephone Number=tel:+00-0000-000000} @{Name=; …

PS /home/RokeJulianLockhart> $Values | ConvertFrom-JSON | Select-Object -ExpandProperty 'Candidate'

Name                    : @{3rd=Surname; 1st=Forename; 2nd=Middle name}
Date of Birth           : +1582-10-15
E-mail Address          : mailTo:address@domain.TLD
Mobile Telephone Number : tel:+00-0000-000000
Blood Group             : 

PS /home/RokeJulianLockhart>
```

</blockQuote>

This isn't desirable — I want to see all of the attributes simultaneously, but with their hierarchies preserved (so merely recursively displaying all of them shan't be useful). As an example:

1.	```PS1
	$Values | ConvertFrom-JSON | Select-Object -ExpandProperty 'Candidate' | ConvertTo-TreeTable
	```
	
1.	<blockQuote>
	
	```log
	Name                    : @{3rd=Surname; 1st=Forename; 2nd=Middle name}
	  3rd                   : Surname
	  1st                   : Forename
	  2nd                   : Middle name
	Date of Birth           : +1582-10-15
	E-mail Address          : mailTo:address@domain.TLD
	Mobile Telephone Number : tel:+00-0000-000000
	Blood Group             :
	```
	
	</blockQiote>

**Comment by rhubarb-geek-nz** [comment]
Have you considered converting it to YAML for output?

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8789685

@rhubarb-geek-nz, no — I've tried to stay away from it due to my experiences with YAML matching https://ruudvanasseldonk.com/2023/01/11/the-yaml-document-from-hell. It just seems like less readable unquoted (https://github.com/json5/json5/blob/de344f0619bda1465a6e25c76f1c0c3dda8108d9/README.md#example:~:text=%7B%0A%20%20%2F%2F%20comments-,unquoted,-%3A%20'and%20you%20can) JSON5. Why do you ask?

**Reply by rhubarb-geek-nz** [reply]
>  Why do you ask?

Because your desired output does not look a million miles away from YAML output. The only difference been a few hyphens and the label value is not in a single column.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8789814

@rhubarb-geek-nz, I see why, then. Unfortunately, the important part of this is to be able to see nested attributes alongside their parents, as a standard fully-expanded GUI tree table would, on demand.

**Comment by rhubarb-geek-nz** [comment]
YAML would be like

```
Candidate:
  Date of Birth: +1582-10-15
  E-mail Address: mailTo:address@domain.TLD
  Mobile Telephone Number: tel:+00-0000-000000
  Name:
    1st: Forename
    2nd: Middle name
    3rd: Surname
Father:
  Date of Birth: +1582-10-15
  E-mail Address: mailTo:address@domain.TLD
  Mobile Telephone Number: tel:+00-0000-000000
  Name:
    1st: Forename
    2nd: Middle name
    3rd: Surname
Mother:
  Date of Birth: +1582-10-15
  E-mail Address: mailTo:address@domain.TLD
  Mobile Telephone Number: tel:+00-0000-000000
  Name:
    1st: Forename
    2nd: Middle name
    3rd: Surname
```

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8790710

Thanks, @rhubarb-geek-nz. Using something https://github.com/cloudbase/powershell-yaml/blob/f89282a327a8cd57ee85c231642db49dfc85de1a/README.md#convertto-yaml to get to that point isn't bad — it's 50% what I want. It's not a prettified table, but it's at least a slightly readable hierarchical key-value pair than unquoted JSON5 is.

**Comment by ThomasNieto** [comment]
The stackoverflow page you referenced doesn't work so not entirely sure what display you want but if you want to display items recursively you can use the `Format-Custom` command with your own custom formatter or the default.

```powershell
$Values | ConvertFrom-JSON | Select-Object -ExpandProperty 'Candidate' | Format-Custom

class PSCustomObject
{
  Name =
    class PSCustomObject
    {
      3rd = Surname
      1st = Forename
      2nd = Middle name
    }
  Date of Birth = +1582-10-15
  E-mail Address = mailTo:address@domain.TLD
  Mobile Telephone Number = tel:+00-0000-000000
  Blood Group =
}
```

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8790940

@ThomasNieto, I mean a hierarchical table, like most file managers provide:

> ![Screenshot_20240314_201824](https://github.com/PowerShell/PowerShell/assets/42837531/f97525cc-d45b-4b4e-8145-74ce98005af5)

**Reply by ThomasNieto** [reply]
The `Microsoft.PowerShell.ConsoleGuiTools` module has a command called `Show-ObjectTree` which will show an interactive menu to navigate objects and nested objects/properties like the file explorer tree view.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8791165

@ThomasNieto, unfortunately, although https://github.com/PowerShell/GraphicalTools/blob/b45f20b63ce2ed8fee13d902df823f5c0057aa65/docs/Microsoft.PowerShell.ConsoleGuiTools/Show-ObjectTree.md#show-objecttree appears promising:

1.	```PS1
	Get-Process | Show-ObjectTree
	```
	
1.	<blockQuote>

	```log    
	Show-ObjectTree: The term 'Show-ObjectTree' is not recognized as a name of a cmdlet, function, script file, or executable program.
	Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
	```

	<blockQuote>

appears to demonstrate that it suffers from the same issue which https://github.com/PowerShell/GraphicalTools/issues/156#issue-1276952302 does. :< Thanks anyway. I'll try it on Windows when I next can. Thanks anyway.

**Reply by ThomasNieto** [reply]
@RokeJulianLockhart Did you install or upgrade the module? It doesn't come with the default install of PowerShell.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8791377

Ah, @ThomasNieto, so it turns out that I'm stupid. I'm so used to having all the tools I could think of already in-built that I rarely ever need to install any modules. Indeed, https://github.com/PowerShell/GraphicalTools/blob/9619e5d1ccd2e2eb76a57ebc869ea21b42c3ecd6/README.md#installation installed it.

However, it doesn't appear to parse the object:

1.	<details>

	```PS1
	$Values = @'
	{
		"Candidate":
		{
			"Name": {
				"3rd": "Surname",
				"1st": "Forename",
				"2nd": "Middle name"
			},
			"Date of Birth": "+1582-10-15",
			"E-mail Address": "mailTo:address@domain.TLD",
			"Mobile Telephone Number": "tel:+00-0000-000000",
			"Blood Group": "" // Unknown
		},
		"Mother":
		{
			"Name": {
				"3rd": "Surname",
				"1st": "Forename",
				"2nd": "Middle name"
			},
			"Date of Birth": "+1582-10-15",
			"E-mail Address": "mailTo:address@domain.TLD",
			"Mobile Telephone Number": "tel:+00-0000-000000",
		},
		"Father":
		{
			"Name": {
				"3rd": "Surname",
				"1st": "Forename",
				"2nd": "Middle name"
			},
			"Date of Birth": "+1582-10-15",
			"E-mail Address": "mailTo:address@domain.TLD",
			"Mobile Telephone Number": "tel:+00-0000-000000",
		},
	}
	'@
	```

1.	*	1.	```PS1
			$Values | Show-ObjectTree
			```
	
		1.	> ![Screenshot_20240316_104948](https://github.com/PowerShell/PowerShell/assets/42837531/6fa61966-63c9-4f6b-b8bf-47f53d4557c3)
		
	*	1.	```PS1
			$Values | ConvertFrom-JSON | Show-ObjectTree
			```
	
		1.	> ![Screenshot_20240316_105413](https://github.com/PowerShell/PowerShell/assets/42837531/3322db31-d9be-4fdd-9e49-bda54d64f0d2)
		1.	When that expansion indicator is invoked, it disappears:
			> ![Screenshot_20240316_105424](https://github.com/PowerShell/PowerShell/assets/42837531/9ccf58b8-3abe-41f7-92b8-e86d5cf89cdb)

**Reply by ThomasNieto** [reply]
Most likely there is a bug with `PSObject` `NoteProperties` not being enumerated. Here is an example with `Get-Process`

![image](https://github.com/PowerShell/PowerShell/assets/38873752/c70db338-a2b5-4aad-bc29-290b0db4caf1)

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8813182

Yeah, @ThomasNieto, I do get the same as you in that case. I'll report a https://github.com/PowerShell/GraphicalTools/issues/new?assignees=&labels=Needs-Triage&projects=&template=Bug_Report.yaml.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8813327

https://github.com/PowerShell/GraphicalTools/issues/242#issue-2190155396

**Comment by roke-julian-lockhart** [comment, marked as answer]
> https://github.com/orgs/PowerShell/discussions/21343#discussion-6373651

Per https://github.com/PowerShell/GraphicalTools/pull/243#issue-2190587377, I can consider https://github.com/orgs/PowerShell/discussions/21343#discussioncomment-8791165 the answer now.
